Initial review of regalloc3

Amanieu opened the pull request:
This PR serves as a placeholder for the initial review of regalloc3.

Amanieu committed:
Initial public release of regalloc3

Amanieu committed:
Move dominator_tree.rs and postorder.rs to debug_utils

These are not used in the main allocation algorithm.

Amanieu committed:
Replace `RegInfo::class_contains` with `RegInfo::class_members`

This allows iterating over all the registers in a class.

Amanieu committed:
Add a cost model to quickly evalute output quality

This is based on LLVM's RegAllocScore and essentailly computes the cost
of all load/store/move/remat operations, multiplied by basic block
frequency.

This allows changes to the register allocation algorithm to be quickly
evaluated for effectiveness.

Amanieu committed:
Avoid panic when parsing a class with no members

Amanieu committed:
Add `SmallEntitySet::map_index` to convert to a different entity type

This is useful when dealing with `RegOrRegGroupSet`.

Amanieu committed:
Collect statistics from the register allocator

The contents of `Stats` are private and not exposed to clients. The only
valid operation that can be done is `Display`ing the stats as a string.

Amanieu committed:
Remove the concept of `SpillSet`s

They are entirely redundant with `ValueSet`, so just use that instead.

Amanieu committed:
Add a move optimization post-processing pass

This can be configured with several optimization levels or even disabled
entirely. The default setting processes blocks in a single pass while
carrying information across block boundaries.

Amanieu committed:
Fix warnings about unused functions

Amanieu committed:
Remove incorrect asserts

These weren't caught earlier because they were only triggered when
trace logging was enabled.

Amanieu committed:
Add dominator information to the `Function` trait

This information is usually already available in clients, so it doesn't
make sense to re-compute it.

Amanieu committed:
Fix clippy warnings and run rustfmt

Amanieu committed:
Remove support for stack maps and safepoints

This has also been removed from regalloc2 in favor of direct support in
Cranelift itself.

Amanieu committed:
Don't allow clobbers to overlap with fixed defs or other clobbers

This was a hack to support regalloc2, but it's better to let the client
handle this case since it simplifies the regalloc logic.

Amanieu committed:
Change `MoveIndex` to be a `u32`

In theory it's possible to have more than 65536 parallel moves due to
block parameters.

Amanieu committed:
Add additional logging when register allocation fails

Amanieu committed:
Optimize coalescing by using a merge instead of a sort

Amanieu committed:
Fix SpillAllocator using data from previous functions

The `sets` field was intentionally not cleared across runs, but this
means that we cannot iterate over it since it may contain stale data.

Amanieu committed:
Fix clippy warning

Amanieu committed:
Don't use a sort to build use lists for each value

Instead, we build a linked list of uses (backed by a vector) and later
append these linked lists to the global use vector once all uses for a
value are known.

This turns out to be faster than collecting uses directly into the
global use vector and sorting it.

Amanieu committed:
Use our own entity types instead of cranelift-entity

The design is mostly derived from cranelift-entity, but several changes
have been made:
- `entity_def` macro instead of `entity_impl` which does more of the
  work and supports non-`u32` index types.
- The `SparseMap` API has been significantly expanded, mostly based on
  `IndexMap`. It no longer uses the `SparseMapValue` trait.
- `SecondaryMap` and `EntitySet` are no longer implicitly resized on
  access. Instead they must be explicitly grown ahead of time. This
  makes it easier to measure the overhead of growing these.
- `CompactList<T>` replaces `EntityList<T>` and holds 2 indices instead
  of just 1, but `T` can be any type instead of an entity reference.
- `SmallEntitySet` is provided in addition to `EntitySet` for small
  fixed-size bitsets.
- Various minor API changes/cleanups.

Various collections used in the allocator have been changed
from `HashMap` or `SecondaryMap` to use `SparseMap` instead which has
O(1) `clear`.

`AllocationUnit` has been turned into a proper entity type and
`AllocationUnit{Map,Set}` are no longer required.

Amanieu committed:
Fix clippy lints

Amanieu committed:
Replace cranelift-bitset with EntitySet

Amanieu committed:
Test arbitrary options in the fuzzer

Amanieu committed:
Optimize bitset scanning when building live range segments

Amanieu committed:
Fix fuzzing for parallel_moves

Amanieu committed:
Add linear live range splitting

This is based on LLVM's local splitting algorithm, but extended to work
across multiple basic blocks.

Amanieu committed:
Move ValueSegment to live_range.rs and add split_at utility function

Amanieu committed:
Track cross-block fixed register hints

Rather than relying on `FixedDef` and `FixedUse` uses for hints, hints
are instead explicitly tracked at the points where a live range would
connect to a fixed register use or def, even across blocks.

This also removes the need to handle empty segments in virtual registers
which were an awkward edge case.

Amanieu committed:
Fix calculation of value live ranges

Amanieu committed:
Fix clippy lints

Amanieu committed:
Fix more clippy links and check doc links in CI

Amanieu committed:
Replace Function::inst_is_terminator with Function::terminator_kind

This explicitly indicates what kind of terminator a block has.

Amanieu committed:
Extend live ranges of operands on return instructions

This reduces the number of live range segments needed to represent a
value if it is live in the block after the return instruction.

Amanieu committed:
Remove `RegOrRegGroup` from public API

This makes the API clearer and more understandable.

Amanieu committed:
Change `RegInfo` and `Function` methods to return iterators of `RegUnit`

This avoids the need to create in-memory slices of `RegUnit` for clients
which map a single `RegUnit` for each `PhysReg`.

Amanieu committed:
Remove `Clone` from `CompactList` and add `swap_remove`

`Clone` is misleading since the clone is shallow: modifications to one
clone could still affect other clones.

Amanieu committed:
Export `entity_def` macro for public use

Amanieu committed:
Change `SmallEntitySet` iterators to copy the set by value

Amanieu committed:
Change `RegUnit` iterators in traits to `Iterator`

These were previously `ExactSizeIterator`

Amanieu committed:
Allow values that are never used

Amanieu committed:
Fix clippy lints

Amanieu committed:
Update to 2024 edition

Amanieu committed:
Optimize interference checking

The previous implementation used iterators which wasn't being optimized
well by LLVM.

Amanieu committed:
Move statistics behind a Cargo feature

Amanieu committed:
Remove `CalleeSaved*` allocation orders

Instead, callee-saved registers should be specified as allocator
constraints on function entry and exit. This allows live range splitting
to potentially move saving and restoring these registers to cold paths.

Amanieu committed:
Don't include stats in the trace log if stats are disabled

Amanieu committed:
Allow critical edge blocks to be explicitly specified

This allows critical edge blocks to contain instructions that expand to
nothing, such as instructions producing undefined values.

Amanieu committed:
Remove unused file

Amanieu committed:
Use incremental interference checking when finding a vreg split

This also switches the backing store for the register matrix to use the
`brie-tree` crate which is significantly faster than `BTreeMap`.

Amanieu committed:
Fix clippy lints

Amanieu committed:
Format and cleanup

Amanieu committed:
Make parsing work with Windows-style newlines

Amanieu committed:
Check `inst_block` in the function validator

Amanieu committed:
Validate that stack_to_stack_class doesn't include spillslots

Amanieu committed:
Fix clippy lints

Amanieu committed:
Minor naming cleanup

Amanieu committed:
Add an adjustment factor to spill weight calculation

This has the effect of "smoothing out" the differences in the length of
short live ranges. In practice this tends to speed up compilation by 10%
while also improving the speed of generated code by about 5%.

Amanieu committed:
Limit `spill_weight_adjust` when fuzzing

Amanieu committed:
Remove allocation order randomization

This turns out to not help compilation speed and was generating worse
code. This also simplifies the `RegInfo` API by removing the concept of
preferred and non-preferred registers in favor of a single allocation
order which is iterated in order.

Amanieu committed:
Fix parallel_moves fuzz test

Amanieu committed:
Track the last allocated register for each `ValueSet` as a hint

This encourages split virtual registers to be allocated to the same
physical register if possible, which can allows more moves to be
eliminated during move optimization.

Amanieu committed:
Use full candidate list when splitting

This allows registers that were hinted but not included in the
allocation order to be considered in the split.

Amanieu committed:
Add additional stats for emergency evictions

Amanieu committed:
Add design document

Amanieu committed:
Merge branch 'empty'

This is done to allow a full review of the main branch in a PR.

cfallin reviewed (commented):
Some initial thoughts on the design doc below. The full 25k lines of code will take me a while to review (sorry!) but I'd be curious on your thoughts on the below in the meantime.

cfallin commented on DESIGN.md:
```diff
@@ -0,0 +1,344 @@
+- Support for register groups, multi-unit registers, explicit stack slots and more flexible register classes.
+- Rematerialization support.
+- Stricter requirements on the input function, which simplifies the allocator implementation:
+  - Block parameters are only allowed on edges with multiple predecessor blocks.
```
> which will be slower in practice

I'd be interested to see data to back that up. In Cranelift at least the redundant-phis pass is a fixpoint loop that needs to build a separate representation (block summaries) and has measurable overhead. Also from a compiler-design point of view, we have never enforced strong normalization on blockparams and these sorts of constraints have a way of making life very tricky for other passes and leading to a long tail of bugs if not enforced at the type level. That's my primary motivation here: I don't want to add complexity to the compiler; this is asking us to kick off a serious refactor in the compiler. Which leads me to...

> It would also add a lot of complexity to the API which currently makes a clear distinction

... which may certainly be true of RA3's design, but isn't really a fundamental argument -- this is only saying that RA3's API was designed around this constraint. RA2 in contrast accepts arbitrary blockparams and does so with a simple, uniform API (every non-returning terminator has block calls, every block call has args, every block has params).

Amanieu commented on DESIGN.md:
```diff
@@ -0,0 +1,344 @@
+
+## Live ranges and uses
+
+The *live range* of a value refers to the paths through a function from its definition to all its uses. Specifically, this covers all instructions across which a value needs to be preserved because it is later used. This is represented as a slice of `ValueSegment`, each covering a linear range between two `LiveRangePoint`.
```
For this section, do you think there is a way I could have worded this to be clearer about points vs ranges?

Amanieu committed:
Improve documentation

Fixes #3

Amanieu committed:
Fix clippy

bjorn3 committed:
Remove ordered-float dependency

bjorn3 committed:
Disable default features of hashbrown

This removes the allocator-api2, equivalent and foldhash dependencies.

bjorn3 committed:
Remove unused bitflags dependency

bjorn3 committed:
Re-enable inline-more feature of hashbrown

Amanieu committed:
Fix incomplete comment in union_find.rs

Amanieu committed:
Work around buggy clippy lint

The `needless_range_loop` lint emits false-positives for
multi-dimensional arrays.

Amanieu committed:
Merge pull request #6 from bjorn3/less_deps

Amanieu committed:
Include instruction number in checker error messages

Amanieu committed:
Properly test usage of blockparams in the first inst of a function

Previously the random function generator would not use any values
defined in the current block in the first instruction of that block,
which unfortunately resulted in some edge case bugs not being caught.

Amanieu committed:
Fix a bug when combining a `Reuse` with a tied operand

This only occurred if in the first instruction of a block, the
source of the reused operand was a blockparam defined in that same
block.

Amanieu committed:
Fix constraint conflict of a blockparam at the start of a block

The separate virtual registers we generate for such conflicts need to
include a `UseKind::BlockparamIn` so that the move resolver knows where
to get incoming values from.

Amanieu committed:
Merge pull request #7 from Amanieu/fix-blockparam

Amanieu committed:
Allow empty classes in RegInfo, but disallow using them in functions

This allows for more flexibility in RegInfo implementations that encode
data into the class index.

Amanieu committed:
Remove the restriction that each `ValueGroup` may only be used once

Amanieu committed:
Add built-in support for multiple entry points

Amanieu committed:
Add fast path in spill cost calculation for uses with zero weight

Amanieu committed:
Fix some logic issues in the move optimizer related to spill slots

Amanieu committed:
Add support for indirect rematerialization

Amanieu committed:
Add a final dead spill elimination pass

Amanieu committed:
Remove Ord from some types where it doesn't make sense

Amanieu committed:
Change default move optimization level to Global

Amanieu committed:
Tweak rematerialization rules in the parallel move resolver

Amanieu committed:
Fix incorrect logic in SpillAllocator::clear

Amanieu committed:
Make fuzz-generated functions respsect MAX_INST_OPERANDS

Amanieu committed:
Merge pull request #8 from Amanieu/remat

Add support for indirect rematerialization

Amanieu committed:
Avoid O(n^2) in debug assertion

Amanieu committed:
Fix clippy warnings
